handler.cpp

```diff
@@ -273,4 +273,156 @@ RequestHandler::Status StatusHandler::HandleRequest(const Request& request, Resp
+    // grow to accommodate the entire line. The growth may be limited by passing
+    // a maximum size to the streambuf constructor.
+    boost::asio::streambuf response;
+    std::unique_ptr<Response> response_ptr;
+
+    // Read the response headers, which are terminated by a blank line.
+    boost::asio::read_until(socket, response, "\r\n\r\n", error);
+    if (error) {
+        std::cout << "Error reading until end of response\n";
+        return nullptr;
+    }
+
+    std::string s( (std::istreambuf_iterator<char>(&response)), std::istreambuf_iterator<char>() );
+    response_ptr = response_parser.Parse(s);
+
+    // Check if there is a "Location" header, signaling redirect
+    std::string redirect_path = "";
+    for (auto &header: response_ptr->GetHeaders()) {
+        if (header.first == "Location") // Get the redirect path
+            redirect_path = header.second;
+    }
+
+    // Handle redirect
+    if ((response_ptr->GetStatus() == 302) && !redirect_path.empty()) {
+        // Find whether rediret path is http or https
+        std::string http_path = redirect_path.substr(1, redirect_path.find(":") - 1);
+        // Assume redirect location has format "http://www.something.com/"
+        redirect_path = redirect_path.substr(redirect_path.find("//") + 2);
+
+        // Find "www.something.com" btween the slashes
+        redirect_path = redirect_path.substr(0, redirect_path.find("/"));
+        
+        // Redirect can be "https" this requires port 443 instead of 80
+        std::string host_ = redirect_path;
+
+        // Set port of https to 443 so it doesn't loop
+        if (http_path == "https")
+            return get_response(path, host_, "443");
+        else    // regular http requests are on port 80
+            return get_response(path, host_, "80");
+    }
+
+    // Read until EOF, writing data to output as we go.
+    while (boost::asio::read(socket, response,
+          boost::asio::transfer_at_least(1), error)){
+            if (error)
+                break;
+        } 
+
+    // Read the rest of the resopnse which will be the response body
+    // Add it with the headers we got before to form a complete response
+    std::string body( (std::istreambuf_iterator<char>(&response)), std::istreambuf_iterator<char>() );
+    std::string whole_response = s + body;
+
+    // Set the response body to only the body, without the headers
+    response_ptr->SetBody(whole_response.substr(whole_response.find("\r\n\r\n") + 4));
+
+    return response_ptr;
+  }
+  catch (std::exception& e)
+  {
+    std::cout << "Exception: " << e.what() << "\n";
+  }
+  return nullptr;
+}
+
+RequestHandler::Status ProxyHandler::HandleRequest(const Request& request, Response* response) {
+    std::cout << "\nProxyHandler::HandleRequest" << std::endl;
+
+    // Pass in the request uri
+    std::string request_uri = request.uri();
+    std::unique_ptr<Response> returned_ptr = get_response(request_uri, m_host, "80");
```

Comment:
why not just replace request_uri with request.uri()? request.uri seems pretty self documenting